### ISSUE TITLE
Bundles vulnerable copy of Expat - please update to 2.2.1

### ISSUE THREAD

**Issue by hartwork** [Stichting-MINIX-Research-Foundation/minix#228]
Hi!

MINIX seems to [bundle *three* outdated vulnerable copy of Expat](https://github.com/search?q=org%3AStichting-MINIX-Research-Foundation+filename%3Axmlparse.c). Please update your copy to version 2.2.1 with the latest security fixes. A change log with details is available at https://github.com/libexpat/libexpat/blob/master/expat/Changes . Thank you!

Best

&nbsp;
Sebastian

**Comment by boricj** [comment]
MINIX 3 imported the NetBSD source tree to get a modern POSIX system a few years ago. The last synchronization happened in 2015 and it's far from being a trivial operation (last time it outright broke the ARM port on hardware and it's still not completely fixed).

So far, synchronization has always been done in one shot (we import a complete NetBSD snapshot). If we start updating components individually, we might end up with a Frankenstein-esque NetBSD mess on our hands in the future. If expat is in that situation, then there's bound to be other obsolete, vulnerable libraries in our source tree which would be fixed by a new synchronization.

@sambuc @dcvmoole what are your opinions on that subject?

**Comment by dcvmoole** [comment]
For security purposes we can selectively upgrade vulnerable components to the new NetBSD version, with the hope that the NetBSD version won't change anymore until the next full resync. In that case, merging later will be relatively easy because our version and the new-NetBSD version are exactly the same. Otherwise, it's going to be more work. However, as a general policy, we cannot wait for resyncs to resolve critical security issues..

**Comment by hartwork** [comment]
> However, as a general policy, we cannot wait for resyncs to resolve critical security issues..

Thanks for saying that.

I think NetBSD has updated Expat to 2.2.1 by now so that part could be synced from them without much effort, maybe.

**Comment by sambuc** [comment]
@hartwork, thank you for the report.

As @dcvmoole mentioned, we make an exception for security patches. Whenever possible we import from our upstream the fixes, if they have been merged there already.

The reason being very simple, in that case, this doesn't introduce any issues in future synchronisation (or imports) from those sources.

I will take a look look into those two, and try to submit a patches for the files in `xsrc`and `minix` the `netbsd` report, however, is a snapshot taken at a specific time, so I don't plan to change it before next code import.

**Comment by hartwork** [comment]
> Whenever possible we import from our upstream the fixes, if they have been merged there already.

If [http://pkgsrc.se/textproc/expat](this) is status quo of NetBSD, they have updated to 2.2.1 already.

**Comment by boricj** [comment]
Oh crap, there's pkgsrc too >_<

Our pkgsrc snapshot of expat (https://github.com/Stichting-MINIX-Research-Foundation/pkgsrc-ng/tree/3.4.0/textproc/expat) is at version 2.2.0. We'll have to update that too and most likely rebuild any packages that rely on it.

**Comment by sambuc** [comment]
The PKGSRC repository is independent from the netbsd repository, so this is basically two different lifecycles.

**Comment by hartwork** [comment]
If PKGSRC is one half, what's the other?

**Comment by sambuc** [comment]
PKGSRC is a repository of software, like gentoo ports, or FreeBSD ports. It is a collection of Makefiles + some patches, to allow installation on divers operating systems of that collection of applications and libraries.

The official location for PKGSRC is http://www.pkgsrc.org, and http://www.pkgsrc.se is a nice interface to look at the available packages, and their descriptions.

**Comment by hartwork** [comment]
So then this is the other half? https://github.com/NetBSD/src

**Comment by sambuc** [comment]
no, the official source repository for NetBSD is still CVS, and available somewhere on http://www.netbsd.org. They have a bugzilla iirc to file in bugs and requests, or you can send a message to one of the mailing lists.

Here are the pointers to those: https://www.netbsd.org/support/

**Comment by hartwork** [comment]
https://github.com/NetBSD/src reads "Automatic conversion of the NetBSD src CVS module, use with care". I read that as "CVS is up to date, already". No?

**Comment by boricj** [comment]
They set up that new, official Git mirror very recently iirc.

**Comment by sambuc** [comment]
CVS is the origin, or reference versioning control system for the NetBSD sources.

The GitHub repository is an automatic conversion. I can not say when it was last updated, or if it is accurate.
